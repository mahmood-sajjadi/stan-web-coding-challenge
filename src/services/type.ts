import type { ProgramType } from "../config";

export interface ListResponse {
    total:   number;
    entries: Entry<ProgramType>[];
}

export interface Entry<T extends ProgramType> {
    title:       string;
    description: string;
    programType: T;
    images:      Images;
    releaseYear: number;
}

export interface Images {
    "Poster Art": PosterArt;
}

export interface PosterArt {
    url:    string;
    width:  number;
    height: number;
}