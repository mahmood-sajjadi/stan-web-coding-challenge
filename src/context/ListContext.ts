import React from 'react';
import type { ListResponse } from '../services/type';

export type Status = 'pending' | 'success' | 'error'

export type ListContextType = {
    status: Status,
    error: string,
    result: ListResponse | undefined,
}

export const ListContext = React.createContext<ListContextType>({
    status: 'pending',
    error: '',
    result: undefined,
});