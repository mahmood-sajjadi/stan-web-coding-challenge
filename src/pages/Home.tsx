import { Link } from "react-router-dom";
import { Item } from "../components/Item";
import { programTypesList } from "../config";
import { List as ListWrapper } from "../components/List";
import { useContext, useMemo } from "react";
import { ListContext } from "../context/ListContext";
import { Loading } from "../components/Loading";

const popularTitleStyle = {
    color: 'black',
}

const posterStyle = {
    backgroundColor: 'black',
    color: 'white',
    lineHeight: '150px',
    fontWeight: 'bold',
    fontSize: '25px',
    textTransform: 'uppercase',
    backgroundSize: 'cover',
}

const linkStyle = {
    textDecoration: 'none',
    width: '100px',
}

const listStyle = {
    gridTemplateColumns: 'repeat(auto-fit, 100px)',
}

export function Home() {
    const list = useContext(ListContext);

    const popularTitles = useMemo(() => {
        return programTypesList.map(programType => (
            <Link to={programType.uri} style={linkStyle} key={programType.type}>
                <Item
                    posterArtUrl={programType.posterArtUrl}
                    title={programType.description}
                    wrapperStyle={popularTitleStyle}
                    posterStyle={posterStyle}
                    className="glass"
                >
                    {programType.title}
                </Item>
            </Link>
        ));
    }, []);

    return (
        <Loading status={list.status}>
            <ListWrapper style={listStyle}>
            {
                popularTitles
            }
            </ListWrapper>
        </Loading>
    )
}