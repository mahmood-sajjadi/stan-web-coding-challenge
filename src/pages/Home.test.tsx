import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom';
import {Home} from './Home'
import { ListContext, type Status } from '../context/ListContext'
import type { ListResponse } from '../services/type'

function testSetup(status: Status, error: string, result: ListResponse) {
    
    return (
        <ListContext.Provider value={{
            status,
            error,
            result,
        }}>
          <MemoryRouter>
            <Home />
          </MemoryRouter>
        </ListContext.Provider>
    )
}

describe('Home', () => {
  it('renders the Home component', () => {
    const home = render(testSetup('success', '', {total: 0, entries: []}));
    expect(home).toMatchSnapshot();
  })

  it('renders the Home component WHEN error', () => {
    const home = render(testSetup('error', 'some error', {total: 0, entries: []}));
    expect(home).toMatchSnapshot();
  })

  it('renders the Home component WHEN pending', () => {
    const home = render(testSetup('pending', '', {total: 0, entries: []}));
    expect(home).toMatchSnapshot();
  })
})