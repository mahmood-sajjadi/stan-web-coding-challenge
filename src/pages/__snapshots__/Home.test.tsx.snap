// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > renders the Home component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: grid; grid-template-columns: repeat(auto-fit, 100px);"
      >
        <a
          data-discover="true"
          href="/series"
          style="text-decoration: none; width: 100px;"
        >
          <div
            style="width: 100px; height: 200px; margin: 0px auto; color: black;"
          >
            <div
              aria-description="Popular Series"
              class="glass"
              style="width: 100%; height: 150px; background-position: center; background-repeat: no-repeat; background-size: cover; text-align: center; background-image: url("/placeholder.png"); background-color: black; color: white; line-height: 150px; font-weight: bold; font-size: 25px; text-transform: uppercase;"
            >
              Series
            </div>
            <span
              style="font-size: 10px;"
            >
              Popular Series
            </span>
          </div>
        </a>
        <a
          data-discover="true"
          href="/movie"
          style="text-decoration: none; width: 100px;"
        >
          <div
            style="width: 100px; height: 200px; margin: 0px auto; color: black;"
          >
            <div
              aria-description="Popular Movies"
              class="glass"
              style="width: 100%; height: 150px; background-position: center; background-repeat: no-repeat; background-size: cover; text-align: center; background-image: url("/placeholder.png"); background-color: black; color: white; line-height: 150px; font-weight: bold; font-size: 25px; text-transform: uppercase;"
            >
              Movies
            </div>
            <span
              style="font-size: 10px;"
            >
              Popular Movies
            </span>
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: grid; grid-template-columns: repeat(auto-fit, 100px);"
    >
      <a
        data-discover="true"
        href="/series"
        style="text-decoration: none; width: 100px;"
      >
        <div
          style="width: 100px; height: 200px; margin: 0px auto; color: black;"
        >
          <div
            aria-description="Popular Series"
            class="glass"
            style="width: 100%; height: 150px; background-position: center; background-repeat: no-repeat; background-size: cover; text-align: center; background-image: url("/placeholder.png"); background-color: black; color: white; line-height: 150px; font-weight: bold; font-size: 25px; text-transform: uppercase;"
          >
            Series
          </div>
          <span
            style="font-size: 10px;"
          >
            Popular Series
          </span>
        </div>
      </a>
      <a
        data-discover="true"
        href="/movie"
        style="text-decoration: none; width: 100px;"
      >
        <div
          style="width: 100px; height: 200px; margin: 0px auto; color: black;"
        >
          <div
            aria-description="Popular Movies"
            class="glass"
            style="width: 100%; height: 150px; background-position: center; background-repeat: no-repeat; background-size: cover; text-align: center; background-image: url("/placeholder.png"); background-color: black; color: white; line-height: 150px; font-weight: bold; font-size: 25px; text-transform: uppercase;"
          >
            Movies
          </div>
          <span
            style="font-size: 10px;"
          >
            Popular Movies
          </span>
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home > renders the Home component WHEN error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Oops, something went wrong...
    </div>
  </body>,
  "container": <div>
    Oops, something went wrong...
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home > renders the Home component WHEN pending 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Loading...
    </div>
  </body>,
  "container": <div>
    Loading...
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
